Uniqueness Typing System

Sushi6006 opened the pull request:
- [x] Code of uniqueness checker is finalised
- [x] In-code docs added
- [ ] Add the docs into *markdown* files

I am not sure if there's anything that I missed. Please let me know if there's anything I can do.

jimbxb commented:
Hey @Sushi6006, nice work on uniqueness checking! :)

One thing you'll need to do is port your changes from `NewParser` over to the `Parser` module. `NewParser` replaced the old parser entirely, so you might need to merge from master again.

There's also a few conflicts that will need to resolve (you can do this with the above merge) before we can merge these changes.

jimbxb reviewed (commented):
Just a few more comments that I think would allow for further modifiers if we ever have any.

Sushi6006 commented on src/NewParser.hs:
```diff
@@ -129,10 +129,16 @@ moduleItemParser v = do
+-- | Process type modifiers.
+processTypeModifiers :: [Ident] -> TypeModifiers
+processTypeModifiers ["unique"] = TypeModifiers True
+processTypeModifiers _          = defaultTypeModifiers
```
Done. Cannot be sure if my code is correct though. I will put up more details in a later commit.

Sushi6006 commented on src/ASTShow.hs:
```diff
@@ -30,6 +30,9 @@ instance Show Module where
             maybeimpl = modImplementation mod
         in " Module " ++ showModSpec (modSpec mod) ++
            bracketList "(" ", " ")" (modParams mod) ++
+           (if tmUniqueness (typeModifiers int)
```
Done. The code has been modified into:
```Haskell
"\n modifiers       : " ++ (show $ typeModifiers int)
```

This time I am slightly more confident that it is correct.

Sushi6006 commented:
I tried my best to merge the files; however there are still quite a few conflicts. Sorry for leaving these to you. Let me know if there is anything I can be helpful with.

Also, great thanks to @jimbxb 's great advises. I have added them into the code base. Hope them works :\

Files not yet updated:
- [AST.hs](src/AST.hs)
  - [ ] LINE 1772 - `data TypeDef`
  - [ ] LINE 3028 - `show`
  - [ ] LINE 3049 - `show`
- [Parser.hs](src/Parser.hs)
  - I am lost in this file. Apologies in advance, but I will leave the pain to you. Or, if someone can briefly explain to me what has changed since "NewParser", I can probably have another go. Do let me know if there is anything I can help though.
  - On LINE 144, I have added `errmsg` for unknown/invalid type modifiers, taking the advice from @jimbxb. However, I can't convince myself that what I have written is correct. Please double-check.

jimbxb commented:
I'll try my best to explain what happened with regards to the parser:
- NewParser was a complete rework of the parser
- Parser (what I'll call old parser) was replaced by NewParser, so the file called NewParser was changed to Parser (effectively removing the old parser)

What you should do from here:
- replace Parser with master's version of Parser
- take your changes made to NewParser from an older commit, and place those in Parser

If you're unsure, feel free to ask any more questions :)

Sushi6006 commented:
> 

That seems fairly easy to understand. I will try to get that done.

jimbxb commented:
I think uniqueness checking needs a little work. I'm not trying to discourage you from adding to this PR, just providing some food for thought.

* Due to uniqueness checking occuring *before* resources have been flattened into arguments, resources aren't checked to ensure uniqueness constraints are met
* Branches might not be handled correctly:
  * A uniquely typed value should be able to be used in all branches of an if block
  * If a condition uses a uniquely typed value and then fails, can the else branch use the value? My instinct says no, so you'd have to ensure that that value is not used in the else branch
* I'm unsure of how semidet code should handle unique types:
  * If a semidet proc fails, can we use any uniquely typed input values again?
* You need to be careful with generics
  * If you input a uniquely typed variable into a generically typed procedure, can the procedure use that value multiple times? If not, how can you ensure this

Some of these issues can probably be remedied by pushing the uniqueness checking further back in the compiler pipeline-- probably somewhere in the bodybuilding stage, but I'm unsure of this, perhaps it needs it's own pass. Pushing it back in the pipeline allows this analysis to be done on LPVM, which is a much more constrainted (and generally easier to reason about) form

Sushi6006 commented:
Thanks for your recommendation. I am definitely willing to solve the issues you have mentioned, but I am not sure if I am capable of doing so at the moment, especially if the implementation goes as far back as you have mentioned.

I have finished adding my work into the newest Parser. I am thinking probably finish with this PR first, and I will try to have a look at the problems you have mentioned. If that is doable, of course.

Also, just to want to mention again that there are still problems unsolved with merging, so the code is not yet tested. I am not 100% sure if they'll still work as they used to/supposed to.

pschachte commented:
I've pulled from this branch and fixed the conflicts that prevented it from passing the tests, and have now merged it as PR #222.  I also partially documented it; I'll document it better once some limitations are removed.  And I extended the error messages so that they include the source position of the error.

Addressing James's comments, branches aren't necessarily a problem.  Unique types need to be linear for each statement sequence, so analysing one branch after another should actually be OK.  Each branch will begin with unique variables not yet read, and each statement that uses one both reads and writes it, so it leaves it in the same state.

However, you're right that there's a problem with unique types in semidet code, because after a failure, variables return to the value they had before the semidet code, which means the value is not unique while the semidet code is executing.  So that's a known bug in this code; I'll put in an issue for it.  You're also right about generics, and I'm not sure how to handle that.  We may need to forbid instantiating a type variable with a unique type; perhaps we can find a way to soften this restriction.  That'll be another issue.

I was thinking about moving the uniqueness checking back to work on LPVM.  That would probably be an easier analysis, and it could perhaps be tied in with the aliasing analysis.  Ultimately, it would be good if a type declared unique could always be destructively updated.  An issue for another time.

But this closes issue #43, which is a valuable improvement.  Thanks for putting in that PR, Terry.
